docs/query-metadata-style-guide.md

```diff
@@ -0,0 +1,152 @@
+
+*   `@name Access to variable in enclosing class`
+*   `@name Array argument size mismatch`
+*   `@name Reference equality test on strings`
+*   `@name Return statement outside function`
+
+
+### Query descriptions `@description`
+
+You must define an `@description` property for your query. This property defines a short help message. Query descriptions should be written as a sentence or short-paragraph of plain prose, with sentence capitalization and full stop. The preferred pattern for alert queries is  `Syntax X causes behavior Y.` Any code elements included in the description should be enclosed in single quotes. For example:
+
+
+*   `@description Using a format string with an incorrect format causes a 'System.FormatException'.`
+*   `@description Commented-out code makes the remaining code more difficult to read.`
+
+### Query ID `@id`
+
+You must specify an `@id` property for your query. It must be unique in the Semmle namespace and should follow the standard Semmle convention. That is, it should begin with the 'language code' for the language that the query analyzes followed by a forward slash. The following language codes are supported:
+
+*   C and C++: `cpp`
+*   C#: `cs`
+*   COBOL: `cobol`
+*   Java: `java`
+*   JavaScript and TypeScript: `js`
+*   Python: `py`
+
+The `@id` should consist of a short noun phrase that identifies the issue that the query highlights. For example:
+
+
+
+*   `@id cs/command-line-injection`
+*   `@id java/string-concatenation-in-loop`
+
+Note, `@id` properties should be consistent for queries that highlight the same issue for different languages. For example, the following queries identify format strings which contain unsanitized input in Java and C++ code respectively:
+
+
+
+*   `@id java/tainted-format-string`
+*   `@id cpp/tainted-format-string`
+
+
+### Query type `@kind`
+
+`@kind` is a required property that defines the type of query. The main query types are:
+
+
+
+*   alerts (`@kind problem`)
+*   alerts containing path information (`@kind path-problem`)
+
+These `@kind` properties support two further mandatory properties which are added by Semmle after the query has been tested, prior to deployment to LGTM. The following information is for reference:
+
+
+
+*   `@precision`–broadly indicates the proportion of query results that are true positives, while also considering their context and relevance:
+    *   `low `
+    *   `medium `
+    *   `high `
+    *   `very high`
+*   `@problem.severity`–defines the level of severity of the alert: 
+    *   `error`–an issue that is likely to cause incorrect program behavior, for example a crash or vulnerability.
+    *   `warning`–an issue that indicates a potential problem in the code, or makes the code fragile if another (unrelated) part of code is changed.
+    *   `recommendation`–an issue that results in code that is hard to read, but is otherwise correct.
```

Comment:
It's probably not all `recommendation`s that are code-is-hard-to-read issues - even though it might be the common case.  So this definition should be loosened up a bit.

Reply:
Changed to: `recommendation`–an issue where the code behaves correctly, but it could be improved.